Submission confirmation email: the time stamp given in the email is incorrect #6101

sznicci opened the pull request:
Fixes #6101 

I haven't seen any straightforward way to add a test to the test suite. Therefore I only tested it manually. (I triggered mail sending on my staging server and checked the mail from a system whose time zone was aligned with the feedback session's time zone.)

<!--
Please follow the naming conventions in the "guidelines for contributing" link above or the pull request may be rejected.

Also, do consider outlining the solution taken as doing so will likely help in the reviewing process
(e.g., when the pull request involves non-trivial changes)
-->

sznicci commented:
Hi,
Could anybody please review this pull request?

damithc commented:
@wkurniawan07 fyi

wkurniawan07 commented on src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:
```diff
@@ -1614,7 +1615,9 @@ public EmailWrapper sendConfirmationEmailForSubmission(String courseId, String f
-            String timestamp = TimeHelper.formatTime12H(Calendar.getInstance().getTime());
+            String timestamp = TimeHelper.formatTime12H(
+                    TimeHelper.convertToUserTimeZone(Calendar.getInstance(TimeZone.getTimeZone("UTC")),
+                            session.getTimeZone()).getTime());
```
The issue is `timestamp` is being passed as `String`. If you can modify such that the timestamp is passed as a `Calendar` or `Date`, you can perform this conversion at the `EmailGenerator` class, thus making the operation testable.

sznicci commented:
Thank you @wkurniawan07 for the review and the advice!
I have updated the pull request and it is ready for review.

I think that it would be better to use `TimeZone` instead of double for the type of the `timeZone` attribute in the `FeedbackSessionAttributes` class. What do you think? I am happy to refactor it if you want me to do it.

sznicci commented:
Hi @wkurniawan07 ,
Could you please look at my changes?
(Sorry, I do not want to be impatient, just the `process.md` document says that we should complete the PR in 2 weeks and it is OK to request a review after 2 days.)

wkurniawan07 commented on src/main/java/teammates/logic/core/EmailGenerator.java:
```diff
@@ -131,18 +132,19 @@ public EmailWrapper generateFeedbackSubmissionConfirmationEmailForInstructor(
             CourseAttributes course, FeedbackSessionAttributes session, String submitUrl,
-            String userName, String userEmail, String timeStamp) {
+            String userName, String userEmail, Calendar timeStamp) {
+        TimeHelper.convertToUserTimeZone(timeStamp, session.getTimeZone());
```
I thought you changed the method such that it did not modify the original parameter? In which case this line does nothing?
Also, while we're at it, you can change the variable name to `timestamp` (small "s") to make it uniform.

wkurniawan07 commented:
Can you also add a test case in `EmailGeneratorTest` where the session timezone is not UTC?

damithc commented:
Let's fix this soon. Users are complaining about the problem.

sznicci commented:
@wkurniawan07 Thank you for the feedback, I am working on it.

@damithc I'll submit a corrected fix today.

sznicci commented:
@wkurniawan07 I corrected what you have requested and it is ready for review.
If there are still more changes needed just let me know and I will fix them asap.

sznicci commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -156,8 +156,50 @@ public static Date getMsOffsetToCurrentTimeInUserTimeZone(int offset, double tim
+        int serverOffset = time.getTimeZone().getRawOffset();
+        int sessionOffset = (int) (60 * 60 * 1000 * timeZone);
+        int offset = sessionOffset - serverOffset;
+        TimeZone sessionTimeZone = getTimeZone(timeZone);
```
@wkurniawan07 Until this line nothing happens, because this solution guarantees that if the method is called twice, then it won't change the timezone again, which would ruin the tests. 
I think this is a nice solution semantically, but if you just want to leave it like it was and adjust the tests accordingly, then I can do it. I am going to do that in another pull request so that you can decide which one do you like better.

wkurniawan07 replied:
The problem with this solution is the effect is more far-reaching than you initially thought. If you do a workspace-wide search for this method (`TimeHelper.convertToUserTimeZone`), you will see that the usage goes much more than just what this issue is about; and they happen to be on areas that are not very well-tested so I am a bit wary if the change is done now.
Let's keep it simple for now? If you want to try implementing the semantically nicer solution, feel free to work on it, and also examine the effects on other parts of the application.

damithc commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -156,8 +156,50 @@ public static Date getMsOffsetToCurrentTimeInUserTimeZone(int offset, double tim
+    }
+    
+    /**
+     * Formats the timestamp to a String with respecting the time zone
```
`respective`?

As this is a public method of a utility class,
1. The header comment should define the precise behavior.
2. Recommended to add unit tests

Same for other new methods in this class.

sznicci commented:
@damithc I have added the requested unit tests and documentation.

@wkurniawan07 I have modified the solution so that it doesn't affect any other part of the project.
I also created a simplified solution, you can find it here : #6153 .
Feel free to use whichever you like.

wkurniawan07 commented:
Closed as #6153 has solved the issue.
